game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/battlegroup/BattleGroup.java

```diff
@@ -0,0 +1,107 @@
+package games.strategy.triplea.delegate.battle.steps.fire.battlegroup;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.engine.data.properties.GameProperties;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Predicate;
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class BattleGroup {
+
+  /** What type of fire type will be used (AA, Normal, etc) */
+  @Builder.Default BattleState.FireType fireType = BattleState.FireType.NORMAL;
+
+  /** Can the offensive casualties return fire? */
+  boolean casualtiesOnOffenseReturnFire;
+  /**
+   * Check the offensive units to determine if the offensive casualties return fire?
+   *
+   * <p>Uses the value of casualtiesOnOffenseReturnFire unless overridden
+   */
+  Predicate<Collection<Unit>> casualtiesOnOffenseReturnFirePredicate;
+  /** Can the defensive casualties return fire? */
+  boolean casualtiesOnDefenseReturnFire;
+  /**
+   * Check the defensive units to determine if the defensive casualties return fire
+   *
+   * <p>Uses the value of casualtiesOnDefenseReturnFire unless overridden
+   */
+  Predicate<Collection<Unit>> casualtiesOnDefenseReturnFirePredicate;
+  /** Squadrons on the offense side */
+  @Builder.Default Collection<FiringSquadron> offenseSquadrons = List.of();
+  /** Squadrons on the defense side */
+  @Builder.Default Collection<FiringSquadron> defenseSquadrons = List.of();
+
+  public static List<BattleGroup> createBattleGroups(final GameData gameData) {
+    return List.of(
+        AaBattleGroups.create(gameData.getUnitTypeList(), gameData.getTechnologyFrontier()),
+        BombardmentBattleGroups.create(gameData.getProperties()),
+        createFirstStrikeBattleGroups(
+            gameData.getUnitTypeList().getAllUnitTypes(), gameData.getProperties()),
+        createGeneralBattleGroups(
+            gameData.getUnitTypeList().getAllUnitTypes(), gameData.getProperties()));
+  }
+
+  private static BattleGroup createFirstStrikeBattleGroups(
+      final Collection<UnitType> unitTypes, final GameProperties properties) {
+    if (Properties.getWW2V2(properties)) {
+      return NormalWW2V2BattleGroups.createFirstStrike(unitTypes, properties);
+    } else {
+      return NormalBattleGroups.createFirstStrike(unitTypes, properties);
+    }
+  }
+
+  private static BattleGroup createGeneralBattleGroups(
+      final Collection<UnitType> unitTypes, final GameProperties properties) {
+    if (Properties.getWW2V2(properties)) {
+      return NormalWW2V2BattleGroups.createGeneral(unitTypes, properties);
+    } else {
+      return NormalBattleGroups.createGeneral(unitTypes, properties);
+    }
+  }
+
+  /**
```

Comment:
Let's have another look at what this class owns and how it is constructed. It is a code smell if a class can be constructed in too many ways. The presence of predicates as part of configuration I think is core to the problem that makes a class too generic and violates DRY and encapsulation (another class constructing this class knows how the predicates are used, it knows what internal data is filtered with it, it knows about the internal data of the class in order to do that, and the filtering logic is spread between the class that constructed this class and this class).